### ISSUE TITLE
bug: lifecycle hooks are not awaited when zone.js has patched Promise constructor

### ISSUE THREAD

**Issue by marcincichocki** [ionic-team/stencil#4459]
### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/stencil/blob/main/CONTRIBUTING.md).
- [X] I agree to follow the [Code of Conduct](https://github.com/ionic-team/stencil/blob/main/CODE_OF_CONDUCT.md).
- [X] I have searched for [existing issues](https://github.com/ionic-team/stencil/issues) that already report this problem, without success.

### Stencil Version

3.2.2

### Current Behavior

Async lifecycle methods (specifically https://tc39.es/ecma262/#prod-AsyncMethod not only methods which return instance of the `Promise`) are not "awaited" due to latest changes: https://github.com/ionic-team/stencil/pull/4250 when zone.js is preset on the page.

Because zone.js is patching `Promise` constructor, but async methods are unaffected (they still return native `Promise` instance) stencil is now comparing instance of a native promise to a patched promise with instanceof check (`nativePromiseFromAsyncMethod instanceof ZoneAwarePromise`). This will always resolve to `false` and lifecycle method will be treated as synchronous, causing all sorts of problems, primarily that render method will not wait for promise to resolve or reject.

### Expected Behavior

Lifecycle methods like `componentWillLoad` and others that can be "awaited" should be if custom promise is used.

### System Info

```shell
System: node 18.16.0
    Platform: darwin (22.5.0)
   CPU Model: Apple M2 Pro (12 cpus)
    Compiler: /Users/marcin.cichocki/dev/playground/dist-bug/node_modules/@stencil/core/compiler/stencil.js
       Build: 1682947516
     Stencil: 3.2.2 🏒
  TypeScript: 4.9.5
      Rollup: 2.42.3
      Parse5: 7.1.2
      Sizzle: 2.42.3
      Terser: 5.17.1
```

### Steps to Reproduce

Run demo with `npm run start` and open console. The order of logs should be 1,2,3, but it is 1,3,2

### Code Reproduction URL

https://github.com/marcincichocki/stencil-promise-issue

### Additional Information

The workaround would be to not use async methods, and return `Promise.resolve()` but that would prevent from using await keyword, which is annoying when working with complex async code (hello promise hell).

Also as a note why we use zone.js: This is a direct dependency of angular, which we use side by side with stencil.

**Comment by alicewriteswrongs** [comment]
hey @marcincichocki thanks for filing this issue! I was just able to confirm that this is not working as intended. The fix is pretty simple, I'm going to open a PR to fix shortly.

**Comment by alicewriteswrongs** [comment]
okie dokie so I just pushed up a tentative fix (#4460) and published a dev build — if you wouldn't mind trying it out to see if it fixes the issue for you:

```sh
npm i @stencil/core@3.3.1-dev.1686078453.7c4f277
```

hopefully that should fix it! 🤞

**Comment by marcincichocki** [comment]
@alicewriteswrongs I tested it and now it works! Thank you for the quick fix.

**Comment by alicewriteswrongs** [comment]
Awesome, glad to hear it worked! Thanks for filing the issue, and I'll ping here when this is merged + released.

**Comment by rwaskiewicz** [comment]
The fix for this issue has been [officially released as a part of Stencil 3.4.0](https://github.com/ionic-team/stencil/releases/tag/v3.4.0). Folks running into this issue should update to v3.4.0 at minimum. I'm going to close this out following the release, but please feel free to open a new issue should this reappear. Thanks!
